EventGenerator/src/CosmicCORSIKA.cc

```diff
@@ -0,0 +1,641 @@
+////////////////////////////////////////////////////////////////////////
+/// \file  CosmicCORSIKA_module.cc
+/// \brief Generator for cosmic-ray secondaries based on pre-generated CORSIKA shower databases.
+///
+/// \author  Matthew.Bass@physics.ox.ac.uk
+////////////////////////////////////////////////////////////////////////
+
+#include "EventGenerator/inc/CosmicCORSIKA.hh"
+
+using CLHEP::Hep3Vector;
+using CLHEP::HepLorentzVector;
+
+namespace mu2e {
+
+  CosmicCORSIKA::CosmicCORSIKA(art::Run &run,
+                              const SimpleConfig &config,
+                              CLHEP::HepRandomEngine &engine)
+      : _refY0(config.getDouble("cosmicCORSIKA.refY0", 0.)), // mm
+        _projectToTargetBox(config.getBool("cosmicCORSIKA.projectToTargetBox", false)),
+        _tOffset(config.getDouble("cosmicCORSIKA.timeOffset", 0.)),
+        _buffer(config.getDouble("cosmicCORSIKA.buffer", 0.)),
+        _showerAreaExtension(config.getDouble("cosmicCORSIKA.showerAreaExtension", 2000.)), // mm
+        _refPointChoice(config.getString("cosmicCORSIKA.refPoint", "UNDEFINED")),
+        _targetBoxXmin(config.getDouble("cosmicCORSIKA.targetBoxXmin", -1000)), // mm
+        _targetBoxXmax(config.getDouble("cosmicCORSIKA.targetBoxXmax", 1000)),  // mm
+        _targetBoxYmin(config.getDouble("cosmicCORSIKA.targetBoxYmin", -1000)), // mm
+        _targetBoxYmax(config.getDouble("cosmicCORSIKA.targetBoxYmax", 1000)),  // mm
+        _targetBoxZmin(config.getDouble("cosmicCORSIKA.targetBoxZmin", -1000)), // mm
+        _targetBoxZmax(config.getDouble("cosmicCORSIKA.targetBoxZmax", 1000)),  // mm
+        _randomXZshift(config.getDouble("cosmicCORSIKA.randomXZshift", 500)), // mm
+        _flat(engine, -_randomXZshift, _randomXZshift),
+        _fluxConstant(config.getDouble("cosmicCORSIKA.fluxConstant", 1.8e4))
+  {
+    config.getVectorString("cosmicCORSIKA.showerInputFiles", _showerInputFiles);
+
+    _showerInputs = _showerInputFiles.size();
+
+    if (_refY0 == 0.)
+      mf::LogInfo("CosmicCORSIKA") << "Using 0. for _refY0!";
+
+    _fileIndex = 0;
+    mf::LogInfo("CosmicCORSIKA") << "Opening: " << _showerInputFiles[_fileIndex] << std::endl;
+    in = fopen(_showerInputFiles[_fileIndex].c_str(), "r");
+    fread(&_garbage, 4, 1, in); // skipping the first (and useless) information
+
+  }
+
```

Comment:
I am not sure I understand this. The list of CORSIKA IDs is not exactly dense (there are missing values, e.g. 4 and from 33 to 49, not sure why). I could use a std::map instead of a std::array though. Also, this is CORSIKA specific, so I don't think it will ever be used outside this module. What do you think?

Reply:
I did not look at it carefully enough to see the missing ones.   If you really think that this is the only place then there is no need to make it a separate function - just be prepared to move it if a someone discovers they need it elsewhere.

If you decide to change the implementation I would still make it a function static const_expr std::arrary - fill the missing slots with an illegal value and test for it.  You don't have to do it this way but I want things like this on everyone's radar when they are designing code.

Reply:
There are many empty slots, I opted for a std::map, I think it's easier to understand and to read.